### ISSUE TITLE
Batched prefill corrupts every row but the first (Falcon3-7B-Instruct)

### ISSUE THREAD

**Issue by kmabeeTT** [tenstorrent/tt-xla#5940]
## Problem

Greedy decoding is a deterministic function of a prompt, so a request's output must not depend on which other requests share its batch. It does: when two or more prefills are scheduled in the same step, **every row except position 0 is corrupted**. Serial prefill is exact.

At concurrency this shows up as runaway repetition — 26/541 ifeval docs loop, versus 1/541 sequentially.

Found while investigating tenstorrent/tt-inference-server#4752 (Falcon3-7B-Instruct accuracy vs the L4 reference), which has the eval-side history. That ticket stays open independently of this one: serialising prefill fixes the loops but recovers only ~2.4 ifeval points, so it does not close the accuracy gap there.

## Repro

Scripts live on the debug branch [`kmabee/falcon3_7b_debug`](https://github.com/tenstorrent/tt-xla/tree/kmabee/falcon3_7b_debug/tools) (commit `d5e3a2cbd`) — this investigation is winding down, so they were not landed on main.

[`tools/batch_invariance_repro.py`](https://github.com/tenstorrent/tt-xla/blob/d5e3a2cbd/tools/batch_invariance_repro.py) — no server, no eval harness. Issues the same prompts one at a time, then together, against one engine instance and diffs the text.

```
# breaks: 8 requests, prefills forced to batch
python tools/batch_invariance_repro.py --bisect --counts 0 --set prefill_batch_threshold=0

# clean: same 8 requests, prefills served one at a time
python tools/batch_invariance_repro.py --bisect --counts 0
```

## Observations

**`prefill_batch_threshold` selects the path, and is not itself buggy** — it routes small prefill bursts to the b1 graph (`scheduler/ascend_scheduler.py:104-112`), so it only hides the defect. Request count is just a proxy for which path runs:

| threshold | requests | prefill | targets matching serial |
|:--|--:|:--|:--|
| 16 (production) | 8, 17 | serial | 8/8 |
| 16 | 20 / 24 / 32 | batched | 2/8 |
| **0** | **8** | **batched** | **2/8** |
| **100000** | **32** | **serial** | **8/8** |

- **Row 0 is immune.** Divergent set is identical in every failing arm: prompt 0 survives, 1-6 break. Points at a per-row offset/index/base-address term that is only correct at zero. Not absolute — with all prompts truncated to a common length the surviving set shifts, so it is positional, not prompt-specific.
- **Deterministic.** Three repeats byte-identical, same divergent set. Not a race.
- **Not ragged-length handling.** Equal-length rows (`--equal-len`) still corrupt 3-4 of 8, so it is not padding-to-the-longest-row.
- **Divergence is early** (char 5 on one prompt), i.e. the batched prefill is already wrong rather than decode drifting. Consistent with first-token top-1 agreement over 541 prompts showing 0 differences — those were single-request prefills, which are the correct path.
- **The arithmetic is fine.** `tools/batch_invariance_probe.py` (JAX, no vLLM) is bit-exact batch-invariant on TT for linear/RMSNorm/MLP/SDPA (uniform, ragged, paged) and the LM head at vocab 131072 with argmax, widths 1-32. CPU is *less* invariant there. So batch-dependent reductions/tiling, paged-KV block layout, and the LM-head reduction are all excluded.
- Prefill and decode use different KV-write ops (`attention_impls/attention.py:473-511`): decode `paged_update_cache`, prefill `paged_fill_cache`. Only prefill is implicated.
- Already-correct, don't re-tread: tail padding rows are sanitised before the KV write (`model_runner.py:1421-1424`); `_get_slot_mapping_metadata` (`model_runner.py:1101`) looks suspicious but is dead code.

**Fixing this does not close the eval gap.** Full ifeval at concurrency 16 with prefill serialised, config otherwise identical to the baseline:

| run | prompt-level strict | loops |
|:--|--:|--:|
| c16 baseline | 63.59 | 26/541 |
| c16, prefill serialised | 65.99 | 2/541 |
| sequential | 74.68 | 1/541 |

The loops are fixed. The score gains only +2.40, which is inside this harness's 2.77-point run-to-run noise, so **a second and larger batch-composition mechanism exists** and serialising prefill does not touch it.

## Next steps

1. **Localise inside the prefill forward.** See the follow-up comment — the error is present in the very first token's *logit value*, so it is in the prefill computation, and part of it is introduced by an `optimization_level=1` pass.
2. **Find mechanism 2.** The offline repro is exactly invariant at 32 concurrent once prefill is serialised, so it does not reach whatever costs the remaining ~8.7 eval points. Likely a regime gap: the repro uses 128-token generations, ctx 2048, no chat template and no KV pressure, while the eval uses long generations, ctx 32768, a chat template and gmu 0.20. Preemption/recompute under KV pressure is the one thing concurrency does that neither sequential nor the repro does — worth trying first.
3. **Workaround meanwhile:** set `prefill_batch_threshold` above peak pending prefills. Restores exact invariance, costs prefill throughput on bursts.

## Reproduced on

| | |
|:--|:--|
| tt-xla | branch `kmabee/falcon3_7b_debug` @ `d5e3a2cbd` (branched from main `7fe709421`) |
| tt-mlir pin | `f209dddc9e` (`third_party/CMakeLists.txt`) |
| tt-metal | `f1f4ff75579` |
| hardware | Blackhole p150b, single device |
| build | Release; `TTMLIR_ENABLE_PERF_TRACE=OFF` (long runs otherwise hit #5504) |

The repro's own defaults are the offline config (ctx 2048, gmu 0.15, `max_num_seqs=32`, `max_num_batched_tokens=1024`). The eval numbers above came from the production serving config: ctx 32768, b32, gmu 0.20, bf16 weights, KV none, `math_fidelity=hifi4`, `fp32_dest_acc_en=True`, `optimization_level=1`, `enable_trace=True`, `cpu_sampling=False`, `prefill_chunk_size=1024`, `min_num_seqs=1`.

**Comment by kmabeeTT** [comment]
## Narrowed: it is the prefill *compute*, and there are two separate defects

Follow-up measurements. These supersede the body's pointer at `paged_fill_cache`: the defect is in the prefill **computation**, not in the KV write at all.

### The error is in the first token's logit, before any KV read-back

Comparing per-step logprob *values* rather than sampled tokens (`--logprob-probe`) moves the picture. For 7 of 8 rows the logprob differs at **step 1** — the token produced by the prefill forward, before any decode step reads the cache — while the sampled token still matches:

```
row 1  step 1: alone -0.044873   batched -0.050093   (same token)
```

The token only flips at step 3 (`--token-sweep`: T=1 and T=2 clean, T=3 breaks row 2, T=8 breaks rows 2 and 4), simply because a ~0.005-0.06 nat error needs a near-tie to change an argmax. So "first token is correct" was true of the token and false of the logit.

This also rules the KV write out directly: on this path prefill attention computes from in-pass q/k/v with no cache read (`attention_impls/attention.py:576-597`; the chunked-prefix and shared-KV branches are both bypassed for a fresh prompt).

### It is position-dependent, and `optimization_level=1` introduces that part

Submitting the **same prompt 8 times** makes every row mathematically identical, so rows can be compared to each other with no cross-run reference (`--same-prompt 8`, prefix caching off):

| | distinct logprobs across 8 identical rows | |
|:--|--:|:--|
| `optimization_level=1` | **2** | row 0 = -0.22772175 (bit-exact to the alone run), rows 1-7 = -0.28869572 |
| `optimization_level=0` | **1** | all rows = -0.29582724 |

A row-independent computation must return 8 identical rows. At opt=1 it does not, so the row index is entering the arithmetic; at opt=0 that term is gone. **The positional defect is introduced by an optimization_level=1 pass.**

### But opt=0 is not a fix — a second, larger effect is opt-independent

At opt=0 the rows agree with each other yet still differ from the single-row run (-0.29583 vs -0.23267), and text still diverges:

| | targets matching the alone run |
|:--|:--|
| batched prefill, opt=1 | 2/8 |
| batched prefill, opt=0 | 4/8 |
| serial prefill, opt=1 | 8/8 |

Deterministic in both cases (repeats byte-identical).

An fp32 CPU golden for that prompt says the batched side is the degraded one, not merely shifted:

| source | first-token logprob | distance from bf16 golden |
|:--|--:|--:|
| HF fp32 CPU (golden) | -0.19526 | — |
| HF bf16 CPU | -0.20117 | — |
| TT single-row / row 0, opt=1 | -0.22772 | 0.027 |
| TT batched rows 1-7, opt=1 | -0.28870 | **0.088** |
| TT batched rows, opt=0 | -0.29583 | **0.095** |

So there are two defects in multi-row prefill:

- **(A) positional** — row 0 correct, rows >= 1 uniformly wrong; introduced by an `optimization_level=1` pass; removed by opt=0.
- **(B) uniform multi-row shift** — present at both opt levels, ~3x further from golden than the single-row path, and on its own still enough to change generated text on half the prompts. This is the larger error and the better candidate for whatever costs the remaining eval points.

### Also excluded

- `torch.ops.tt.scaled_dot_product_attention` is row-invariant at prefill shape (identical rows, B=8, seq 128, causal — bit-exact, `tools/sdpa_row_invariance.py`). The earlier op probe only covered decode (q_len=1), so this closes that gap.
- `generate_attn_mask` (`model_runner.py:165`) builds a single `[1,1,L,S]` mask from `context_lens[0]` only and looks like an obvious culprit, but it is **dead code** — the live copy is `pooling_runner.py:118`, used only by pooling models. The generative path passes `attn_mask=None` and masks causally inside the kernel, and causality keeps row padding out of real positions.
- `logits_indices` is correct per row (`num_scheduled_tokens_per_req - 1`).
- `torch.ops.tt.sparse_matmul` is MoE-only; unused by dense Falcon3.

### Next

1. Bisect the `optimization_level=1` passes to name the one that adds defect (A).
2. Characterise (B) — it is opt-independent, so likely structural to the multi-row prefill graph rather than a pass. Per-layer intermediate comparison between a 1-row and an N-row prefill of the same prompt would localise it.

Repro for all of the above: `tools/batch_invariance_repro.py` on `kmabee/falcon3_7b_debug` @ `d5e3a2cbd` with `--same-prompt`, `--logprob-probe`, `--token-sweep`.
